### ISSUE TITLE
.gitignore should be avoided during build.

### ISSUE THREAD

**Issue by manas-maji** [google-github-actions/deploy-cloudrun#468]
### TL;DR

As with typescript project "dist" or "build" folder generally ignored from remote repo. But including dist folder in .gitignore deployment gets failed.

### Expected behavior

It should not consider .gitignore during build as it is working locally with `docker build` command.

### Observed behavior

Build gets failed after including dist in .gitignore.

### Action YAML

```yaml
name: Deploy to Cloud Run

on:
  push:
    branches:
      - master
    paths:
      - "src/**"
      - "Dockerfile"
      - ".github/workflows/deploy.yaml"
      - ".gitignore"

env:
  PROJECT_ID: project_id 
  SERVICE: service_name
  REGION: region

jobs:
  deploy:
    permissions:
      contents: 'read'
      id-token: 'write'

    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Build to JavaScript
        run: npm install && npm run build   

      - name: Google Auth
        id: auth
        uses: 'google-github-actions/auth@v1'
        with:
          workload_identity_provider: '${{ secrets.GCP_WIF_PROVIDER }}' 
          service_account: '${{ secrets.GCP_WIF_SERVICE_ACCOUNT }}' 

      - name: Deploy to Cloud Run
        id: deploy
        uses: google-github-actions/deploy-cloudrun@v1
        with:
          project_id: ${{ env.PROJECT_ID }}
          region: ${{ env.REGION }}
          service: ${{ env.SERVICE }}
          source: ./
          env_vars: |
            NODE_ENV=production

      - name: Cloud Run URL
        run: echo ${{ steps.deploy.outputs.url }}
```

### Log output

```text
2023-10-02T14:52:33.9872166Z Requested labels: ubuntu-latest
2023-10-02T14:52:33.9872414Z Job defined at: manas-maji/***/.github/workflows/deploy.yaml@refs/heads/master
2023-10-02T14:52:33.9872582Z Waiting for a runner to pick up this job...
2023-10-02T14:52:34.7269455Z Job is waiting for a hosted runner to come online.
2023-10-02T14:52:38.9607993Z Job is about to start running on the hosted runner: GitHub Actions 2 (hosted)
2023-10-02T14:52:43.0773575Z Current runner version: '2.309.0'
2023-10-02T14:52:43.0804243Z ##[group]Operating System
2023-10-02T14:52:43.0805012Z Ubuntu
2023-10-02T14:52:43.0805395Z 22.04.3
2023-10-02T14:52:43.0805782Z LTS
2023-10-02T14:52:43.0806167Z ##[endgroup]
2023-10-02T14:52:43.0806653Z ##[group]Runner Image
2023-10-02T14:52:43.0807112Z Image: ubuntu-22.04
2023-10-02T14:52:43.0807476Z Version: 20230924.1.0
2023-10-02T14:52:43.0808135Z Included Software: https://github.com/actions/runner-images/blob/ubuntu22/20230924.1/images/linux/Ubuntu2204-Readme.md
2023-10-02T14:52:43.0808930Z Image Release: https://github.com/actions/runner-images/releases/tag/ubuntu22%2F20230924.1
2023-10-02T14:52:43.0809563Z ##[endgroup]
2023-10-02T14:52:43.0810025Z ##[group]Runner Image Provisioner
2023-10-02T14:52:43.0810494Z 2.0.310.1
2023-10-02T14:52:43.0810890Z ##[endgroup]
2023-10-02T14:52:43.0811712Z ##[group]GITHUB_TOKEN Permissions
2023-10-02T14:52:43.0813657Z Contents: read
2023-10-02T14:52:43.0814120Z Metadata: read
2023-10-02T14:52:43.0814956Z ##[endgroup]
2023-10-02T14:52:43.0819131Z Secret source: Actions
2023-10-02T14:52:43.0819751Z Prepare workflow directory
2023-10-02T14:52:43.1756460Z Prepare all required actions
2023-10-02T14:52:43.1980128Z Getting action download info
2023-10-02T14:52:43.5174913Z Download action repository 'actions/checkout@v3' (SHA:f43a0e5ff2bd294095638e18286ca9a3d1956744)
2023-10-02T14:52:44.1797620Z Download action repository 'google-github-actions/auth@v1' (SHA:35b0e87d162680511bf346c299f71c9c5c379033)
2023-10-02T14:52:44.6719717Z Download action repository 'google-github-actions/deploy-cloudrun@v1' (SHA:e62f655d5754bec48078a72edc015367b01ee97b)
2023-10-02T14:52:45.3602016Z Complete job name: deploy
2023-10-02T14:52:45.4841646Z ##[group]Run actions/checkout@v3
2023-10-02T14:52:45.4842026Z with:
2023-10-02T14:52:45.4842359Z   repository: manas-maji/***
2023-10-02T14:52:45.4843218Z   token: ***
2023-10-02T14:52:45.4843498Z   ssh-strict: true
2023-10-02T14:52:45.4843801Z   persist-credentials: true
2023-10-02T14:52:45.4844107Z   clean: true
2023-10-02T14:52:45.4844414Z   sparse-checkout-cone-mode: true
2023-10-02T14:52:45.4844739Z   fetch-depth: 1
2023-10-02T14:52:45.4845011Z   fetch-tags: false
2023-10-02T14:52:45.4845275Z   lfs: false
2023-10-02T14:52:45.4845541Z   submodules: false
2023-10-02T14:52:45.4845820Z   set-safe-directory: true
2023-10-02T14:52:45.4846105Z env:
2023-10-02T14:52:45.4846377Z   PROJECT_ID: ***
2023-10-02T14:52:45.4846678Z   SERVICE: ***
2023-10-02T14:52:45.4846963Z   REGION: ***
2023-10-02T14:52:45.4847228Z ##[endgroup]
2023-10-02T14:52:45.7945546Z Syncing repository: manas-maji/***
2023-10-02T14:52:45.7947620Z ##[group]Getting Git version info
2023-10-02T14:52:45.7948302Z Working directory is '/home/runner/work/***/***'
2023-10-02T14:52:45.7950074Z [command]/usr/bin/git version
2023-10-02T14:52:45.8038971Z git version 2.42.0
2023-10-02T14:52:45.8111080Z ##[endgroup]
2023-10-02T14:52:45.8127028Z Temporarily overriding HOME='/home/runner/work/_temp/aab6788f-b3ad-44a7-a1c6-05eb0419141a' before making global git config changes
2023-10-02T14:52:45.8127732Z Adding repository directory to the temporary git global config as a safe directory
2023-10-02T14:52:45.8128566Z [command]/usr/bin/git config --global --add safe.directory /home/runner/work/***/***
2023-10-02T14:52:45.8159055Z Deleting the contents of '/home/runner/work/***/***'
2023-10-02T14:52:45.8164392Z ##[group]Initializing the repository
2023-10-02T14:52:45.8169287Z [command]/usr/bin/git init /home/runner/work/***/***
2023-10-02T14:52:45.8259133Z hint: Using 'master' as the name for the initial branch. This default branch name
2023-10-02T14:52:45.8260780Z hint: is subject to change. To configure the initial branch name to use in all
2023-10-02T14:52:45.8261411Z hint: of your new repositories, which will suppress this warning, call:
2023-10-02T14:52:45.8261799Z hint: 
2023-10-02T14:52:45.8262729Z hint: 	git config --global init.defaultBranch <name>
2023-10-02T14:52:45.8263070Z hint: 
2023-10-02T14:52:45.8263542Z hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
2023-10-02T14:52:45.8264146Z hint: 'development'. The just-created branch can be renamed via this command:
2023-10-02T14:52:45.8264529Z hint: 
2023-10-02T14:52:45.8264855Z hint: 	git branch -m <name>
2023-10-02T14:52:45.8285822Z Initialized empty Git repository in /home/runner/work/***/***/.git/
2023-10-02T14:52:45.8301680Z [command]/usr/bin/git remote add origin https://github.com/manas-maji/***
2023-10-02T14:52:45.8353066Z ##[endgroup]
2023-10-02T14:52:45.8353829Z ##[group]Disabling automatic garbage collection
2023-10-02T14:52:45.8359306Z [command]/usr/bin/git config --local gc.auto 0
2023-10-02T14:52:45.8400382Z ##[endgroup]
2023-10-02T14:52:45.8401293Z ##[group]Setting up auth
2023-10-02T14:52:45.8408524Z [command]/usr/bin/git config --local --name-only --get-regexp core\.sshCommand
2023-10-02T14:52:45.8452806Z [command]/usr/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'core\.sshCommand' && git config --local --unset-all 'core.sshCommand' || :"
2023-10-02T14:52:45.8918360Z [command]/usr/bin/git config --local --name-only --get-regexp http\.https\:\/\/github\.com\/\.extraheader
2023-10-02T14:52:45.8961193Z [command]/usr/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :"
2023-10-02T14:52:45.9240879Z [command]/usr/bin/git config --local http.https://github.com/.extraheader AUTHORIZATION: basic ***
2023-10-02T14:52:45.9286401Z ##[endgroup]
2023-10-02T14:52:45.9287216Z ##[group]Fetching the repository
2023-10-02T14:52:45.9297697Z [command]/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=1 origin +1ca7a2164ebc2e85d5c8c30abde01a1b8bddfdb1:refs/remotes/origin/master
2023-10-02T14:52:46.3846415Z remote: Enumerating objects: 16, done.        
2023-10-02T14:52:46.3847201Z remote: Counting objects:   6% (1/16)        
2023-10-02T14:52:46.3847862Z remote: Counting objects:  12% (2/16)        
2023-10-02T14:52:46.3848497Z remote: Counting objects:  18% (3/16)        
2023-10-02T14:52:46.3849124Z remote: Counting objects:  25% (4/16)        
2023-10-02T14:52:46.3849734Z remote: Counting objects:  31% (5/16)        
2023-10-02T14:52:46.3850355Z remote: Counting objects:  37% (6/16)        
2023-10-02T14:52:46.3850986Z remote: Counting objects:  43% (7/16)        
2023-10-02T14:52:46.3851636Z remote: Counting objects:  50% (8/16)        
2023-10-02T14:52:46.3852268Z remote: Counting objects:  56% (9/16)        
2023-10-02T14:52:46.3852906Z remote: Counting objects:  62% (10/16)        
2023-10-02T14:52:46.3853542Z remote: Counting objects:  68% (11/16)        
2023-10-02T14:52:46.3854154Z remote: Counting objects:  75% (12/16)        
2023-10-02T14:52:46.3854786Z remote: Counting objects:  81% (13/16)        
2023-10-02T14:52:46.3858196Z remote: Counting objects:  87% (14/16)        
2023-10-02T14:52:46.3862455Z remote: Counting objects:  93% (15/16)        
2023-10-02T14:52:46.3864043Z remote: Counting objects: 100% (16/16)        
2023-10-02T14:52:46.3864958Z remote: Counting objects: 100% (16/16), done.        
2023-10-02T14:52:46.3865916Z remote: Compressing objects:   7% (1/13)        
2023-10-02T14:52:46.3866589Z remote: Compressing objects:  15% (2/13)        
2023-10-02T14:52:46.3867489Z remote: Compressing objects:  23% (3/13)        
2023-10-02T14:52:46.3868640Z remote: Compressing objects:  30% (4/13)        
2023-10-02T14:52:46.3869323Z remote: Compressing objects:  38% (5/13)        
2023-10-02T14:52:46.3870479Z remote: Compressing objects:  46% (6/13)        
2023-10-02T14:52:46.3871399Z remote: Compressing objects:  53% (7/13)        
2023-10-02T14:52:46.3872739Z remote: Compressing objects:  61% (8/13)        
2023-10-02T14:52:46.3874187Z remote: Compressing objects:  69% (9/13)        
2023-10-02T14:52:46.3875106Z remote: Compressing objects:  76% (10/13)        
2023-10-02T14:52:46.3876253Z remote: Compressing objects:  84% (11/13)        
2023-10-02T14:52:46.3876924Z remote: Compressing objects:  92% (12/13)        
2023-10-02T14:52:46.3877578Z remote: Compressing objects: 100% (13/13)        
2023-10-02T14:52:46.3878520Z remote: Compressing objects: 100% (13/13), done.        
2023-10-02T14:52:46.4044037Z remote: Total 16 (delta 0), reused 13 (delta 0), pack-reused 0        
2023-10-02T14:52:46.4125331Z From https://github.com/manas-maji/***
2023-10-02T14:52:46.4126028Z  * [new ref]         1ca7a2164ebc2e85d5c8c30abde01a1b8bddfdb1 -> origin/master
2023-10-02T14:52:46.4163002Z ##[endgroup]
2023-10-02T14:52:46.4163568Z ##[group]Determining the checkout info
2023-10-02T14:52:46.4164083Z ##[endgroup]
2023-10-02T14:52:46.4164581Z ##[group]Checking out the ref
2023-10-02T14:52:46.4167946Z [command]/usr/bin/git checkout --progress --force -B master refs/remotes/origin/master
2023-10-02T14:52:46.4241866Z Reset branch 'master'
2023-10-02T14:52:46.4247294Z branch 'master' set up to track 'origin/master'.
2023-10-02T14:52:46.4253235Z ##[endgroup]
2023-10-02T14:52:46.4309575Z [command]/usr/bin/git log -1 --format='%H'
2023-10-02T14:52:46.4348677Z '1ca7a2164ebc2e85d5c8c30abde01a1b8bddfdb1'
2023-10-02T14:52:46.4674463Z ##[group]Run npm install && npm run build
2023-10-02T14:52:46.4674927Z [36;1mnpm install && npm run build[0m
2023-10-02T14:52:46.4742316Z shell: /usr/bin/bash -e {0}
2023-10-02T14:52:46.4742646Z env:
2023-10-02T14:52:46.4742939Z   PROJECT_ID: ***
2023-10-02T14:52:46.4743252Z   SERVICE: ***
2023-10-02T14:52:46.4743525Z   REGION: ***
2023-10-02T14:52:46.4743794Z ##[endgroup]
2023-10-02T14:52:51.3372974Z 
2023-10-02T14:52:51.3374114Z added 271 packages, and audited 272 packages in 4s
2023-10-02T14:52:51.3374871Z 
2023-10-02T14:52:51.3376590Z 26 packages are looking for funding
2023-10-02T14:52:51.3377160Z   run `npm fund` for details
2023-10-02T14:52:51.3391796Z 
2023-10-02T14:52:51.3392139Z found 0 vulnerabilities
2023-10-02T14:52:51.7425663Z 
2023-10-02T14:52:51.7426735Z > ***@1.0.0 build
2023-10-02T14:52:51.7427327Z > tsc
2023-10-02T14:52:51.7427710Z 
2023-10-02T14:52:54.2456998Z ##[group]Run google-github-actions/auth@v1
2023-10-02T14:52:54.2457276Z with:
2023-10-02T14:52:54.2458222Z   workload_identity_provider: ***
2023-10-02T14:52:54.2458693Z   service_account: ***
2023-10-02T14:52:54.2459105Z   create_credentials_file: true
2023-10-02T14:52:54.2459454Z   export_environment_variables: true
2023-10-02T14:52:54.2459799Z   cleanup_credentials: true
2023-10-02T14:52:54.2460140Z   access_token_lifetime: 3600s
2023-10-02T14:52:54.2460839Z   access_token_scopes: https://www.googleapis.com/auth/cloud-platform
2023-10-02T14:52:54.2461230Z   retries: 3
2023-10-02T14:52:54.2461550Z   backoff: 250
2023-10-02T14:52:54.2461913Z   id_token_include_email: false
2023-10-02T14:52:54.2462321Z env:
2023-10-02T14:52:54.2462652Z   PROJECT_ID: ***
2023-10-02T14:52:54.2463014Z   SERVICE: ***
2023-10-02T14:52:54.2463301Z   REGION: ***
2023-10-02T14:52:54.2463636Z ##[endgroup]
2023-10-02T14:52:54.6648452Z Created credentials file at "/home/runner/work/***/***/gha-creds-511ee3d7d02fc0c8.json"
2023-10-02T14:52:54.6817438Z ##[group]Run google-github-actions/deploy-cloudrun@v1
2023-10-02T14:52:54.6817922Z with:
2023-10-02T14:52:54.6818290Z   project_id: ***
2023-10-02T14:52:54.6818878Z   region: ***
2023-10-02T14:52:54.6819224Z   service: ***
2023-10-02T14:52:54.6819581Z   source: ./
2023-10-02T14:52:54.6819939Z   env_vars: NODE_ENV=production

2023-10-02T14:52:54.6952548Z   skip_default_labels: false
2023-10-02T14:52:54.6952959Z   no_traffic: false
2023-10-02T14:52:54.6953229Z env:
2023-10-02T14:52:54.6953511Z   PROJECT_ID: ***
2023-10-02T14:52:54.6953811Z   SERVICE: ***
2023-10-02T14:52:54.6954404Z   REGION: ***
2023-10-02T14:52:54.6954927Z   CLOUDSDK_AUTH_CREDENTIAL_FILE_OVERRIDE: /home/runner/work/***/***/gha-creds-511ee3d7d02fc0c8.json
2023-10-02T14:52:54.6955661Z   GOOGLE_APPLICATION_CREDENTIALS: /home/runner/work/***/***/gha-creds-511ee3d7d02fc0c8.json
2023-10-02T14:52:54.6956424Z   GOOGLE_GHA_CREDS_PATH: /home/runner/work/***/***/gha-creds-511ee3d7d02fc0c8.json
2023-10-02T14:52:54.6956930Z   CLOUDSDK_CORE_PROJECT: ***
2023-10-02T14:52:54.6957283Z   CLOUDSDK_PROJECT: ***
2023-10-02T14:52:54.6957614Z   GCLOUD_PROJECT: ***
2023-10-02T14:52:54.6957923Z   GCP_PROJECT: ***
2023-10-02T14:52:54.6958256Z   GOOGLE_CLOUD_PROJECT: ***
2023-10-02T14:52:54.6958564Z ##[endgroup]
2023-10-02T14:52:56.1103931Z [command]/usr/bin/tar xz --warning=no-unknown-keyword --overwrite -C /home/runner/work/_temp/892548d8-9344-4c31-863b-1642402fbb1d -f /home/runner/work/_temp/b5f534aa-fb9f-4997-8b49-f3f36f4002b2
2023-10-02T14:53:15.5706616Z Successfully authenticated
2023-10-02T14:53:15.5708457Z Running: gcloud run deploy *** --quiet --source ./ --update-env-vars NODE_ENV=production --update-labels managed-by=github-actions,commit-sha=1ca7a2164ebc2e85d5c8c30abde01a1b8bddfdb1 --platform managed --format json --region *** --project ***
2023-10-02T14:53:56.6957730Z ##[error]google-github-actions/deploy-cloudrun failed with: failed to execute gcloud command `gcloud run deploy *** --quiet --source ./ --update-env-vars NODE_ENV=production --update-labels managed-by=github-actions,commit-sha=1ca7a2164ebc2e85d5c8c30abde01a1b8bddfdb1 --platform managed --format json --region *** --project ***`: This command is equivalent to running `gcloud builds submit --tag [IMAGE] ./` and `gcloud run deploy *** --image [IMAGE]`

Building using Dockerfile and deploying container to Cloud Run service [***] in project [***] region [***]
Building and deploying...
Uploading sources..........done
Building Container................................................................................................................................................................................................................................................................................................................................................................................failed
Deployment failed
ERROR: (gcloud.run.deploy) Build failed; check build logs for details
2023-10-02T14:53:56.7064339Z Post job cleanup.
2023-10-02T14:53:56.7975828Z Removed exported credentials at "/home/runner/work/***/***/gha-creds-511ee3d7d02fc0c8.json".
2023-10-02T14:53:56.8079613Z Post job cleanup.
2023-10-02T14:53:56.9315032Z [command]/usr/bin/git version
2023-10-02T14:53:56.9376068Z git version 2.42.0
2023-10-02T14:53:56.9443094Z Temporarily overriding HOME='/home/runner/work/_temp/56708637-acac-46ff-ba58-9250555e5c32' before making global git config changes
2023-10-02T14:53:56.9444652Z Adding repository directory to the temporary git global config as a safe directory
2023-10-02T14:53:56.9451105Z [command]/usr/bin/git config --global --add safe.directory /home/runner/work/***/***
2023-10-02T14:53:56.9508710Z [command]/usr/bin/git config --local --name-only --get-regexp core\.sshCommand
2023-10-02T14:53:56.9572603Z [command]/usr/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'core\.sshCommand' && git config --local --unset-all 'core.sshCommand' || :"
2023-10-02T14:53:56.9890820Z [command]/usr/bin/git config --local --name-only --get-regexp http\.https\:\/\/github\.com\/\.extraheader
2023-10-02T14:53:56.9924667Z http.https://github.com/.extraheader
2023-10-02T14:53:56.9940831Z [command]/usr/bin/git config --local --unset-all http.https://github.com/.extraheader
2023-10-02T14:53:56.9987062Z [command]/usr/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :"
2023-10-02T14:53:57.0561637Z Cleaning up orphan processes
```

### Additional information

Cloud build log.

starting build "020e316d-a0f9-4383-bf08-cc2b027aa777"

FETCHSOURCE
Fetching storage object: gs://***_cloudbuild/source/1696258398.817477-5ace1a31b9e9437a8dfeed6d3a79cbb5.tgz#1696258399325755
Copying gs://***_cloudbuild/source/1696258398.817477-5ace1a31b9e9437a8dfeed6d3a79cbb5.tgz#1696258399325755...
/ [0 files][    0.0 B/ 39.1 KiB]                                                
/ [1 files][ 39.1 KiB/ 39.1 KiB]                                                
Operation completed over 1 objects/39.1 KiB.                                     
BUILD
Already have image (with digest): gcr.io/cloud-builders/docker
Sending build context to Docker daemon  145.4kB

Step 1/7 : FROM node:20-alpine
20-alpine: Pulling from library/node
96526aa774ef: Pulling fs layer
bd866a276fb2: Pulling fs layer
86562bfdb018: Pulling fs layer
20bf5151d902: Pulling fs layer
20bf5151d902: Waiting
86562bfdb018: Verifying Checksum
86562bfdb018: Download complete
96526aa774ef: Verifying Checksum
96526aa774ef: Download complete
20bf5151d902: Verifying Checksum
20bf5151d902: Download complete
96526aa774ef: Pull complete
bd866a276fb2: Verifying Checksum
bd866a276fb2: Download complete
bd866a276fb2: Pull complete
86562bfdb018: Pull complete
20bf5151d902: Pull complete
Digest: sha256:37750e51d61bef92165b2e29a77da4277ba0777258446b7a9c99511f119db096
Status: Downloaded newer image for node:20-alpine
 ---> e722ab357037
Step 2/7 : WORKDIR /app
 ---> Running in 553f633e3fa0
Removing intermediate container 553f633e3fa0
 ---> becdf2cb43e2
Step 3/7 : COPY package*.json ./
 ---> 8636bd4874a2
Step 4/7 : RUN npm install --only=production
 ---> Running in a37fd71731f3
npm WARN config only Use `--omit=dev` to omit dev dependencies from the install.

added 207 packages, and audited 208 packages in 7s

21 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
Removing intermediate container a37fd71731f3
 ---> 4a353c0217e0
Step 5/7 : COPY dist/* .
COPY failed: no source files were specified
ERROR
ERROR: build step 0 "gcr.io/cloud-builders/docker" failed: step exited with non-zero status: 1

**Comment by sethvargo** [comment]
What does your Dockerfile look like? If you run Docker locally, it will have access to the `dist/` folder. If you build remotely with the `dist/` folder git-ignored, then no folder exists and the build will fail.

**Comment by manas-maji** [comment]
> What does your Dockerfile look like? If you run Docker locally, it will have access to the `dist/` folder. If you build remotely with the `dist/` folder git-ignored, then no folder exists and the build will fail.

Thanks for your reply @sethvargo 

Here is my Dockerfile.

```
FROM node:20-alpine

WORKDIR /app

COPY package*.json ./
RUN npm install --only=production

COPY dist/* ./

EXPOSE 8080
CMD [ "node", "index.js" ]
```

I have also checked `dist/` folder existence after step [2](https://github.com/google-github-actions/deploy-cloudrun/issues/468#:~:text=%2D%20name%3A%20Build%20to%20JavaScript%0A%20%20%20%20%20%20%20%20run%3A%20npm%20install%20%26%26%20npm%20run%20build) with `ls -R` command. It exists with all required files inside during workflow run.

Let me know if I missed something.

**Comment by sethvargo** [comment]
Sorry, I don't see where you're building a Docker container image in the GitHub Action YAML you provided above.

**Comment by manas-maji** [comment]
I'm using deployment from source. You can check source argument [here](https://github.com/google-github-actions/deploy-cloudrun/issues/468#:~:text=%24%7B%7B%20env.SERVICE%20%7D%7D-,source%3A%20./,-env_vars%3A%20%7C).

Let me try to explain more clearly with the following screenshot from GitHub Actions logs output.

![image](https://github.com/google-github-actions/deploy-cloudrun/assets/59717976/86a6ef81-6bc7-4ccd-995c-ef9ff209aa3c)

As you can see after checkout step there is no `dist` directory. But after build to JavaScript step we have that directory. So `dist` directory is present before cloud run deploy step runs. But [this](https://github.com/google-github-actions/deploy-cloudrun/issues/468#:~:text=install%20%2D%2Donly%3Dproduction-,COPY%20dist/*%20./,-EXPOSE%208080%0ACMD) is getting failed during Cloud build. [Here](https://github.com/google-github-actions/deploy-cloudrun/issues/468#:~:text=Step%205/7%20%3A%20COPY%20dist/*%20.%0ACOPY%20failed%3A%20no%20source%20files%20were%20specified) is the log from Cloud build.

If I remove `dist` directory from my `.gitignore` file then it's working without any issue. But we generally don't want to push `dist` directory to remote repository, that's why I putted it in `.gitignore` file and bad things (deployment failure) start.

Let me know if it's clear now.

**Comment by sethvargo** [comment]
Assuming you _don't_ want to commit the `dist/` into source, then you have the following options:

1. Build the container locally in GitHub Actions, push it to GCR, and deploy that _image_ to Cloud Run (use the `image` field instead of `source`). You're basically already doing this, but you aren't pushing the image. In essence, you're building twice - once on GitHub and then again on Cloud Build.

2. Update the Dockerfile to do the `npm build`. If you move the `npm build` step into the Dockerfile (ideally as a multi-stage build to reduce size), then the `dist/` folder will be created and stored in the container. This is arguably the best solution, since it ensures the source OS matches the destination OS.

**Comment by manas-maji** [comment]
> Assuming you _don't_ want to commit the `dist/` into source, then you have the following options:
> 
> 1. Build the container locally in GitHub Actions, push it to GCR, and deploy that _image_ to Cloud Run (use the `image` field instead of `source`). You're basically already doing this, but you aren't pushing the image. In essence, you're building twice - once on GitHub and then again on Cloud Build.
> 2. Update the Dockerfile to do the `npm build`. If you move the `npm build` step into the Dockerfile (ideally as a multi-stage build to reduce size), then the `dist/` folder will be created and stored in the container. This is arguably the best solution, since it ensures the source OS matches the destination OS.

Yes, I'm familiar with the options you mentioned, and I've actually used Option 1 for a different project. As for Option 2, personally, I'm not inclined towards it because it introduces unnecessary files into the image.

However, I'm perplexed about why including the `dist` directory in the `.gitignore` file is causing Cloud Build to fail during deployment from source. Any insights into why this might be the case?

**Comment by sethvargo** [comment]
Do you have a .gcloudignore? What happens if you run `gcloud run deploy --source` directly?

**Comment by manas-maji** [comment]
> Do you have a .gcloudignore? What happens if you run `gcloud run deploy --source` directly?

No, I didn't initially have a `.gcloudignore` file. However, I created one by duplicating the contents of the `.gitignore` file and removing the `dist` folder entry. This approach worked perfectly.

### Here's how the source deployment process operates:

1. **Checking Ignore Files:**
   - The deployment process first looks for a `.gcloudignore` file in the source directory.
   - If `.gcloudignore` is not found, it then checks for a `.gitignore` file.

2. **Ignoring Specified Files:**
   - The system ignores the files and directories listed in either the `.gcloudignore` or `.gitignore` file.

3. **Copying to Cloud Storage:**
   - It then copies all remaining contents from the source directory to cloud storage, excluding the files mentioned in the ignore file.

4. **Building and Deploying:**
   - Finally, the deployment proceeds by building and deploying the application using the remaining files.



@sethvargo Thanks for your valuable time. Now it's pretty much clear to me.
